lib/rspec/rails/matchers/have_reported_error.rb

```diff
@@ -0,0 +1,272 @@
+        # @param expected_error_or_message [Class, String, Regexp, nil]
+        #   Error class, message string, or message pattern
+        # @param expected_message [String, Regexp, nil]
+        #   Expected message when first param is a class
+        def initialize(expected_error_or_message, expected_message)
+          @attributes = {}
+          @warn_about_nil_error = expected_error_or_message.nil?
+
+          case expected_error_or_message
+          when UNDEFINED
+            @expected_error = nil
+            @expected_message = expected_message
+          when String, Regexp
+            @expected_error = nil
+            @expected_message = expected_error_or_message
+          else
+            @expected_error = expected_error_or_message
+            @expected_message = expected_message
+          end
+        end
+
+        def with_context(expected_attributes)
+          conflicting_keys = @attributes.keys & expected_attributes.keys
+          unless conflicting_keys.empty?
+            raise ArgumentError, "Attribute keys #{conflicting_keys.inspect} are already defined. " \
+                                "Chaining with_context calls should not overwrite existing attributes. " \
+                                "Use a single with_context call with all attributes instead."
+          end
+          @attributes.merge!(expected_attributes)
+          self
+        end
+
+        def and(_)
+          raise ArgumentError, "Chaining is not supported"
+        end
+
+        def matches?(block)
+          warn_about_nil_error! if @warn_about_nil_error
+
+          @reports = ErrorCollector.record do
+            block.call
+          end
+
+          return false if @reports.empty?
+          return false unless error_matches_expectation?
+
+          return attributes_match_if_specified?
+        end
+
+        def supports_block_expectations?
+          true
+        end
+
+        def supports_value_expectations?
+          false
+        end
+
+        def description
+          base_desc = if @expected_error
+                        "report a #{@expected_error} error"
+                      else
+                        "report an error"
+                      end
+
+          message_desc = if @expected_message
+                           case @expected_message
+                           when Regexp
+                             " with message matching #{@expected_message}"
+                           when String
+                             " with message '#{@expected_message}'"
+                           end
+                         else
+                           ""
+                         end
+
+          attributes_desc = @attributes.empty? ? "" : " with #{@attributes}"
+
+          base_desc + message_desc + attributes_desc
+        end
+
+        def failure_message
+          if !@reports.empty? && !@attributes.empty?
+            report_context = @reports.last.context
```

Comment:
`last` is incorrect if several errors could have been reported.

Reply:
If there were no errors reported matching by class/message, even if the expectation has a `with_context` qualifier, don't even mention context attributes in the failure message.

But if **some** reports did match by class/message (or **ALL** if the expectation did not specify class/message/regexp), list those errors with their attributes.

Reply:
`matching_reports` (plural) should do better